internal/jobcompleter/job_completer.go

```diff
@@ -56,170 +60,468 @@ type InlineJobCompleter struct {
+	c.BaseStartStop.Stop()
+	c.asyncCompleter.Stop()
 }
 
-// As configued, total time from initial attempt is ~7 seconds (1 + 2 + 4) (not
-// including jitter). I put in a basic retry algorithm to hold us over, but we
-// may want to rethink these numbers and strategy.
-const numRetries = 3
+func (c *BatchCompleter) Subscribe(subscribeFunc func(update CompleterJobUpdated)) {
+	c.withSubscribe.Subscribe(subscribeFunc)
+	c.asyncCompleter.Subscribe(subscribeFunc)
+}
 
-func withRetries(c *baseservice.BaseService, f func(ctx context.Context) (*rivertype.JobRow, error)) (*rivertype.JobRow, error) { //nolint:varnamelen
-	retrySecondsWithoutJitter := func(attempt int) float64 {
-		// Uses a different algorithm (2 ** N) compared to retry policies (4 **
-		// N) so we can get more retries sooner: 1, 2, 4, 8
-		return math.Pow(2, float64(attempt))
+func (c *BatchCompleter) WaitStarted() <-chan struct{} {
+	return c.started
+}
+
+func (c *BatchCompleter) waitOrInitBacklogChannel(ctx context.Context) {
+	c.setStateParamsMu.RLock()
+	var (
+		backlogSize = len(c.setStateParams)
+		waitChan    = c.waitOnBacklogChan
+		waiting     = c.waitOnBacklogWaiting
+	)
+	c.setStateParamsMu.RUnlock()
+
+	if waiting {
+		<-waitChan
+		return
 	}
 
-	retrySeconds := func(attempt int) float64 {
-		retrySeconds := retrySecondsWithoutJitter(attempt)
+	// Not at max backlog. A little raciness is allowed here: multiple
+	// goroutines may have acquired the read lock above and seen a size under
+	// limit, but with all allowed to continue it could put the backlog over its
+	// maximum. The backlog will only be nominally over because generally max
+	// backlog >> max workers, so consider this okay.
+	if backlogSize < c.maxBacklog {
+		return
+	}
 
-		// Jitter number of seconds +/- 10%.
-		retrySeconds += retrySeconds * (c.Rand.Float64()*0.2 - 0.1)
+	c.setStateParamsMu.Lock()
+	defer c.setStateParamsMu.Unlock()
 
-		return retrySeconds
+	// Check once more if another process has already started waiting (it's
+	// possible for multiple to race between the acquiring the lock above). If
+	// so, we fall through and allow this insertion to happen, even though it
+	// might bring the batch slightly over limit, because arranging the locks
+	// otherwise would get complicated.
+	if c.waitOnBacklogWaiting {
+		return
 	}
 
-	tryOnce := func() (*rivertype.JobRow, error) {
-		ctx := context.Background()
+	// Tell all future insertions to start waiting. This one is allowed to fall
+	// through and succeed even though it may bring the batch a little over
+	// limit.
+	c.waitOnBacklogChan = make(chan struct{})
+	c.waitOnBacklogWaiting = true
+	c.Logger.WarnContext(ctx, c.Name+": Hit maximum backlog; completions will wait until below threshold", "max_backlog", c.maxBacklog)
+}
 
-		ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
-		defer cancel()
+// As configued, total time from initial attempt is ~7 seconds (1 + 2 + 4) (not
+// including jitter).
+const numRetries = 3
```

Comment:
Ugh, so this is really tricky to get right. It's 7 seconds, but if the loop is timing out every time (timeouts being a likely failure candidate), the total time is more like 37 seconds (7 seconds + 3 * 10 seconds), which is a really long time. I modified the comment above to be more explicit about that case.

This can be bad because the completer can still block a client shutdown, and blocking it for this length of time could be a serious problem. I've actually seen this happen to slightly smaller extents when a bad failure happens in the completer running a benchmark. I'll hit Ctrl+C to stop it, and it takes ~10 seconds to exit.

I think I'm going to punt on more changes on this in favor of a separate project thinking more about these retries. I'm not convinced that we're doing the right thing at all when it comes to this stuff. I think retrying once or _maybe_ twice makes a certain amount of sense, but given these are likely indicative of something very degenerate happening in the DB, continuing to beat a dead horse by retrying over and over again is likely to lead to more failures and more unnecessary database contention. Even the retry schedule is probably wrong — the first retry should probably be happening even sooner after an initial failure because in the event of a truly intermittent failure, even 1s is quite a long time.